### ISSUE TITLE
[Regression][SM100/SM103] FlashInfer TRTLLM NVFP4 MoE tile-192 path produces NaNs

### ISSUE THREAD

**Issue by zianglih** [sgl-project/sglang#34629]
## Summary

`flashinfer_trtllm` online NVFP4 MoE becomes numerically invalid after upgrading the aligned FlashInfer trio beyond `0.6.16rc4`. On Blackwell, the newly exposed tile-192 TRTLLM_GEN BMM path produces non-finite MoE output; the registered 200-example GSM8K test consequently scores `0.0`.

This is a dependency/integration regression, not a feature-specific regression in the SGLang change where it was first noticed:

- the same pre-rebase SGLang checkout fails with FlashInfer `0.6.16.post4` and `0.6.17`;
- that exact checkout passes when only the FlashInfer trio is changed to `0.6.15.post1` or `0.6.16rc4`;
- forcing a valid tile-128 tactic passes the full correctness workload, while both shape-valid tile-192 tactics fail during startup autotuning before any serving request.

## Public CI symptom (B200)

The registered test fails publicly on four NVIDIA B200 GPUs at SGLang `8531fe0a8ca4e3b6d92b4134be399aaac261c636`:

- Job: https://github.com/sgl-project/sglang/actions/runs/31607763933/job/94151038518
- Driver: `580.126.09`
- PyTorch: `2.13.0+cu130`
- FlashInfer: `flashinfer-python==0.6.17`, `flashinfer-cubin==0.6.17`, `flashinfer-jit-cache==0.6.17+cu130`
- First `M=896` prefill: `Assertion 'NaN detected! sampler: next_token_logits' failed.` on all four ranks
- Result: GSM8K score `0.000`

The B200 job independently corroborates the same first-`M=896`, all-rank NaN symptom on SM100. Its later CUDA abort/coredumps are downstream effects and are not separately root-caused here.

## B300 environment

```text
Hardware: 8x NVIDIA B300 SXM6 AC (SM103, 267.7 GiB, 148 SMs); server uses TP4/EP4
Driver: 590.48.01
Container: lmsysorg/sglang:nightly-dev-cu13-20260812-c7c03ec5
amd64 digest: sha256:e948f2d4adb7d3ca8336d855f667ab26ab27bc7515702336e3c6f7093121bf23
OS: Ubuntu 24.04.4 LTS, Linux 6.8.0-137-generic, glibc 2.39
Python: 3.12.3
PyTorch / CUDA: 2.13.0+cu130 / 13.0
NCCL: 2.29.7
SGLang package-bisect SHA: 3f3002d4a88d964f2f19ef501755f994cd5d25cc
Also reproduced at clean current SHA: 8531fe0a8ca4e3b6d92b4134be399aaac261c636
Model: Qwen/Qwen3-30B-A3B-Instruct-2507-FP8
MoE backend: flashinfer_trtllm
Quantization: nvfp4_online
Workload: GSM8K completion API, 200 examples, 128 threads, max_tokens=512
```

The environment held constant across the package bisect:

```text
FLASHINFER_NVFP4_4OVER6=1
FLASHINFER_NVFP4_4OVER6_ERR_MODE=MSE
FLASHINFER_NVFP4_4OVER6_ERR_USE_FAST_MATH=1
FLASHINFER_NVFP4_4OVER6_E4M3_USE_256=1
SGLANG_FP4_IGNORED_LAYERS=shared_expert,model.layers.40,model.layers.41,model.layers.42,model.layers.43,model.layers.44,model.layers.45,model.layers.46,model.layers.47
SGLANG_ENABLE_ASYNC_ASSERT=true
```

SGLang logs `--disable-shared-experts-fusion is automatically set`; shared-expert fusion is already disabled and is not the failing path.

## Reproduction

The model requires Hugging Face authentication. Install SGLang from the tested checkout, then install an aligned FlashInfer trio. Use a fresh server and cache for each version.

```bash
git clone https://github.com/zianglih/sglang.git
cd sglang
git checkout 3f3002d4a88d964f2f19ef501755f994cd5d25cc
python3 -m pip install -e python

VERSION=0.6.17
python3 -m pip uninstall -y \
  flashinfer-python flashinfer-cubin flashinfer-jit-cache
rm -rf /root/.cache/flashinfer
python3 -m pip install --no-cache-dir --no-deps \
  "flashinfer-python==${VERSION}"
python3 -m pip install --no-cache-dir --no-deps \
  --index-url https://flashinfer.ai/whl \
  "flashinfer-cubin==${VERSION}"
python3 -m pip install --no-cache-dir --no-deps \
  --index-url https://flashinfer.ai/whl/cu130 \
  "flashinfer-jit-cache==${VERSION}+cu130"
python3 -m pip check
python3 -m flashinfer show-config

cd test
SGLANG_ENABLE_ASYNC_ASSERT=true \
python3 registered/backends/test_flashinfer_nvfp4_online_moe_backend.py \
  TestFlashinferTrtllmGenMoeBackendNvFp4Online.test_gsm8k -f
```

The test launches:

```text
sglang serve \
  --model-path Qwen/Qwen3-30B-A3B-Instruct-2507-FP8 \
  --moe-runner-backend flashinfer_trtllm \
  --cuda-graph-max-bs-decode 128 \
  --tp-size 4 --ep-size 4 \
  --quantization nvfp4_online \
  --mem-fraction-static 0.7
```

Repeat with `VERSION=0.6.15.post1`, `0.6.16rc4`, and `0.6.16.post4`. Keep `flashinfer-python`, `flashinfer-cubin`, and `flashinfer-jit-cache` aligned; the JIT cache uses the matching `+cu130` build.

## Package-only bisect on B300

Same SGLang SHA (`3f3002d4...`), image, model, arguments, environment, hardware, and full 200-example workload in every row:

| Aligned FlashInfer trio | Result | GSM8K score | Latency | Throughput | First observed failure |
|---|---:|---:|---:|---:|---|
| `0.6.15.post1` | PASS | `0.960` | `4.2688 s` | `6747.30 tok/s` | none |
| `0.6.16rc4` | PASS | `0.945` | `4.7273 s` | `6153.64 tok/s` | none |
| `0.6.16.post4` | FAIL | `0.000` | `147.96 s` | n/a | pure prefill, `M=1792`; NaNs on all four ranks |
| `0.6.17` | FAIL | `0.000` | n/a | n/a | first concurrent prefill; reduced trace reaches `M=896`; NaNs on all four ranks |

This brackets the measured runtime regression after `0.6.16rc4`. FlashInfer source/tag inspection points to the `0.6.16rc5` line as the first carrying the corresponding TRTLLM_GEN BMM artifact/launcher update, but rc5 itself has not yet been run.

## Isolation matrix on 0.6.17 / B300

| Single change | Scope | Result |
|---|---|---|
| default | 1 example, 1 thread, `max_tokens=512` | PASS |
| default | full 200/128/512 | FAIL, score `0.000` |
| `--disable-prefill-cuda-graph` | full 200/128/512 | FAIL |
| Triton attention + page size 64 | full 200/128/512 | FAIL |
| `flashinfer_trtllm_routed` | reduced concurrent run | FAIL |
| deferred finalize disabled | reduced concurrent run | FAIL |
| `SGLANG_TRTLLM_MOE_PDL_MAX_TOKENS=0` | reduced concurrent run | FAIL |
| `--enforce-disable-flashinfer-allreduce-fusion` | full 200/128/512 | FAIL |
| `SGLANG_FORCE_COARSE_WAR_BARRIER=1` | full 200/128/512 | FAIL |
| `--disable-flashinfer-autotune` | full 200/128/512 | FAIL |
| force exact `[128,19]` through `M=16384`, v0.6.17 cubins | full 200/128/512, prefill graph disabled | PASS: score `0.955`, no non-finite marker or async assertion |
| same `[128,19]`, substitute exact same-name v0.6.15 FC1+FC2 cubins | 24 requests / 24 threads / `max_tokens=1`, prefill graph disabled | PASS |
| force `[192,1]` or `[192,3]`, v0.6.17 cubins | startup autotune, `M=4096` | FAIL: non-finite finalized MoE output on all ranks |
| device sync at FP4 entry; stream sync after routing, FC1, explicit quantization, or FC2 | reduced eager run, one probe per run | FAIL at every boundary |
| `--disable-overlap-schedule` | full 200/128/512 | PASS, score `0.955` |
| `CUDA_LAUNCH_BLOCKING=1` | full 200/128/512 | PASS, score `0.940` |

The attention backend and prefill CUDA graph are not required. PDL, routed-vs-FromLogits dispatch, deferred finalize, FlashInfer all-reduce fusion, and SGLang's coarse WAR barrier are insufficient explanations. Since tile 192 becomes non-finite during startup autotuning, SGLang request overlap/concurrency is not required; serialization merely prevents/hides the default serving failure.

## Root-cause status

Confirmed:

- This is FlashInfer-version dependent: package-only replacement flips pass/fail on identical SGLang code.
- It reproduces on B200/SM100 and B300/SM103.
- Layer 0 enters the FlashInfer FP4 MoE op with finite input, block scales, per-token scales, and router data, then receives non-finite finalized BF16 output.
- It is unrelated to the CuTe DSL feature where the failure was first encountered: the same pre-rebase tree fails, `fp4_utils.py` is not on this runtime activation path, and the test uses `flashinfer_trtllm`.
- Synchronizing at FP4 entry and after each high-level internal boundary (routing, FC1, explicit quantization, FC2) does not make the result finite.

Strong current FlashInfer lead, not yet a final kernel root cause:

- The last passing line pins `b368d003.../batched_gemm-da58956-b4ac80e`; failing versions pin the regenerated `5988e15c.../batched_gemm-3d40263-3e19f0a` artifact.
- The artifact update first appears on the release branch in [FlashInfer #4252](https://github.com/flashinfer-ai/flashinfer/pull/4252) and on the v0.6.17 line in [#4180](https://github.com/flashinfer-ai/flashinfer/pull/4180).
- The corrupt serving invocation is an aggregated MoE batch (`M=15605` in the finite trace), not the nearby scheduler line that reports 896 new tokens.
- Forced tile-128 `[128,19]` passes the full correctness test with the v0.6.17 cubins (score `0.955`); the exact same-name v0.6.15 FC1+FC2 pair also stays finite in the reduced crash reproducer.
- Tile 192 has only two shape-valid tactics for `I=768`: `[192,1]` and `[192,3]`. Both fail during the `M=4096` dummy-model forward used by startup autotuning. They change FC1 but share FC2 raw config 787, making that FC2 or its tile-192 integration contract the strongest localization. The finalized-output probe does not yet prove the first corrupt internal tensor.

FlashInfer issue with the artifact/tactic analysis and package bisect: https://github.com/flashinfer-ai/flashinfer/issues/4486

## Mitigation vs. fix

Verified mitigations:

- pin the aligned FlashInfer trio to `0.6.15.post1` or tested `0.6.16rc4`; or
- suppress tile 192 / force a valid tile-128 tactic for this model shape; or
- run with `--disable-overlap-schedule` as a serving workaround.

`CUDA_LAUNCH_BLOCKING=1` is diagnostic only. The root fix belongs in FlashInfer's tile-192 BMM path; SGLang should also avoid accepting a numerically corrupt tactic and add an integration regression test.

No exact SGLang duplicate was found for this version/shape/NaN regression.

**Comment by zianglih** [comment]
### Root-cause correction: FlashInfer tile-192 tactic, not SGLang overlap

I corrected the issue body after completing the tactic A/B.

- Layer 0 enters FlashInfer FP4 MoE with finite input/scales/router data and receives non-finite finalized BF16 output.
- Forced tile-128 `[128,19]` through the 16K profile passes the full 200-example/128-thread GSM8K test with FlashInfer `0.6.17`: score `0.955`, latency `11.087 s`, output throughput `2580.322 tok/s`, no non-finite marker, and no async assertion.
- Both shape-valid tile-192 tactics, `[192,1]` and `[192,3]`, produce non-finite output during the `M=4096` startup-autotune forward, before serving begins. They use different FC1 cubins but share the only shape-valid tile-192 FC2, raw config 787.
- Device synchronization at FP4 entry and stream synchronization after routing, FC1, explicit quantization, and FC2 do not fix it.

Therefore request overlap can expose/mask the default path but is not required for reproduction; the defect is in FlashInfer's v0.6.16rc5+ tile-192 TRTLLM_GEN BMM path. The SGLang-side mitigation is to avoid tile 192 (or pin FlashInfer to the tested `0.6.16rc4` line), plus reject numerically corrupt tactics during autotuning. The corresponding FlashInfer report is https://github.com/flashinfer-ai/flashinfer/issues/4486.

**Comment by SSHdotCodes** [comment]
Same failure mode on **SM120 (RTX PRO 6000 Blackwell)** with FlashInfer `0.6.17`, adding a second architecture to this report.

The presentation differs: instead of non-finite output during startup autotune, the selected tactic here passes warmup and then **silently corrupts serving output**. Decode emits a few correct tokens and then collapses into a run of token id `0` (`!` in this tokenizer) until `max_tokens`:

```
To understand how a modern Mixture-of-Experts (Mo!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!...
```

### Environment

- NVIDIA RTX PRO 6000 Blackwell, SM120, 96 GB, driver `595.84`
- CUDA 13.0.3, PyTorch `2.13.0+cu130`, `flashinfer-python==0.6.17`
- Image `lmsysorg/sglang@sha256:12d3392bdc8be8d35e9a95f191df6aef99c5114bdbefd41bfdc7e760e6d25ec1`
- Model `RadixArk/Qwen3.8-Flash-Next-NVFP4` (512 experts, top-10, NVFP4 W4A4)
- Resolved `moe_runner_backend=flashinfer_cutlass` — SGLang logs *"trtllm-gen MoE kernels are SM100-only"* on SM120 — but the autotuned custom ops are still `trtllm::fused_moe::gemm1/gemm2`.

### Selected tactics

From `.../sglang/flashinfer/autotune/0.6.17/sm120/<key>/rank_tp0_pp0_dp0.json`, identical across restarts:

```json
"('trtllm::fused_moe::gemm1', 'MoERunner', ((1, 2560), (512, 1280, 160), (0,), (512, 2560, 40), (0,)), ())": ["MoERunner", 17],
"('trtllm::fused_moe::gemm2', 'MoERunner', ((1, 2560), (512, 1280, 160), (0,), (512, 2560, 40), (0,)), ())": ["MoERunner", 57]
```

### Controlled A/B

1024-token greedy generations, single request, counting generations that collapse to token id 0:

| config | corrupted | decode tok/s |
|---|---|---|
| autotune on (gemm1=17, gemm2=57) | **36/36**, and **18/18** on a second start | 104.9 |
| `--flashinfer-autotune-skip-ops trtllm::fused_moe::gemm1` | 0/15 | 98.4 |
| `--disable-flashinfer-autotune` | **0/36** | 96.6 |

Skipping `gemm1` alone is sufficient at this shape, i.e. gemm1 tactic 17 is the offender while gemm2 tactic 57 is fine here. On the configuration we shipped (which adds NEXTN/MTP, so `gemm2` is additionally tuned at the extra verify shapes) autotune had to be off entirely: 0 corrupted across 100+ long generations on 4 separate starts.

### The tactic id space is shape-dependent

Hand-writing `gemm2=57` into the cache for every tuned shape makes startup fail outright:

```
Check failed: (id2 >= mGemm1TacticCount && id2 < mGemm1TacticCount + mGemm2TacticCount)
    is false: Invalid gemm2 profile id: 57
```

So an id that is valid at one shape can be out of range at another. Since the cache is keyed on the tuned shape and replayed on neighbouring shapes, the silent-corruption case looks like that same mismatch without a bounds check — consistent with @zianglih's finding that the failing tile-192 tactics were shape-valid but numerically wrong.

### Detection note, possibly relevant for CI

This is easy to miss with short probes. On the server measuring **36/36** corrupt at 1024 output tokens, 256-token requests were **0/48 clean**. Any correctness gate under a few hundred output tokens will not see it.

Config, benchmark harness and raw measurements: https://github.com/SSHdotCodes/qwen-3.8-flash-next-pro6000

Happy to run tactic sweeps on SM120 if that would help narrow which tile configs are affected.

**Comment by zianglih** [comment]
Solving by:
- https://github.com/sgl-project/sglang/pull/36954

**Comment by zianglih** [comment]
Fixed, affected ci re enabled: https://github.com/sgl-project/sglang/pull/36985

**Comment by SSHdotCodes** [comment]
Re-tested after the FlashInfer 0.6.18 bump (#36954) that closed this: **the SM120 invalid-tactic corruption I reported above is reduced but not fixed.**

Setup: RTX PRO 6000 Blackwell Workstation Edition (SM120), `RadixArk/Qwen3.8-Flash-Next-NVFP4`, `--moe-runner-backend flashinfer_cutlass`, NEXTN MTP (steps 3 / topk 1 / draft 4). Base image pinned at nightly-dev `d91c3682` with `pip install --no-deps flashinfer-python==0.6.18 flashinfer-cubin==0.6.18 "flashinfer-jit-cache==0.6.18+cu130"`. Corruption metric: 1024-token generations, fail = output collapses into a `!!!!` run (token id 0). Same box, same flags, only the autotune flag varies:

| config | corrupt | median decode tok/s |
|---|---|---|
| 0.6.17, autotune on (my earlier report) | 36/36 | ~105 |
| **0.6.18, autotune on** | **5/36 (13.9%)** | 173.5 |
| 0.6.18, `--disable-flashinfer-autotune` | 0/36 | 158.4 |

Fisher exact for 5/36 vs 0/36 is ~0.027, so this is not noise. 0.6.18 evidently fixes the tile-192 BMM regression this issue tracked, but the per-shape `trtllm::fused_moe::gemm1/gemm2` tactic-id aliasing from my earlier comment still bites on SM120: tactic ids are numbered per problem shape, the cache replays them on neighbouring shapes, and ranking is latency-only. (`Invalid gemm2 profile id` when hand-pinning across shapes is the same defect caught by a bounds check; the silent 5/36 is it not being caught.)

`--disable-flashinfer-autotune` remains the only clean mitigation here (~9% decode cost on this workload). Happy to open a separate SM120 issue and/or run candidate fixes on this card.

**Comment by SSHdotCodes** [comment]
Follow-up: I ran a deeper bisection of the residual 0.6.18 corruption on SM120 (tactic-level numerical sweeps, graph-replay soaks, PDL-off, dedicated-workspace, and an in-server output-verification guard). The tactics themselves verify numerically correct in isolation and in eager serving paths; the corruption needs non-default fused-MoE tactics executing inside the decode CUDA graphs, is content-anchored, and the tuner's per-boot tactic lottery explains the flaky rates. Full elimination matrix, harnesses, and a serving repro are now in flashinfer-ai/flashinfer#4841. Mitigation remains --disable-flashinfer-autotune (measured speed-equivalent to skipping just the two fused-MoE ops on this model).
